cdc/redo/writer/file.go

```diff
@@ -0,0 +1,533 @@
+type Writer struct {
+	cfg *FileWriterConfig
+	op  *writerOptions
+	// maxCommitTS is the max commitTS among the events in one log file
+	maxCommitTS atomic.Uint64
+	// the ts used in file name
+	commitTS atomic.Uint64
+	// the ts send with the event
+	eventCommitTS atomic.Uint64
+	state         atomic.Uint32
+	gcRunning     atomic.Bool
+	size          int64
+	file          *os.File
+	bw            *pioutil.PageWriter
+	uint64buf     []byte
+	storage       storage.ExternalStorage
+	sync.RWMutex
+}
+
+// NewWriter ... TODO: extract to a common rotate Writer
+func NewWriter(ctx context.Context, cfg *FileWriterConfig, opts ...Option) *Writer {
+	if cfg == nil {
+		log.Panic("FileWriterConfig can not be nil")
+		return nil
+	}
+
+	if cfg.FlushIntervalInMs == 0 {
+		cfg.FlushIntervalInMs = defaultFlushIntervalInMs
+	}
+	cfg.MaxLogSize *= megabyte
+	if cfg.MaxLogSize == 0 {
+		cfg.MaxLogSize = defaultMaxLogSize
+	}
+	if cfg.S3Storage {
+		if cfg.S3URI == nil {
+			log.Panic("S3URI can not be nil",
+				zap.String("change feed", cfg.ChangeFeedID))
+		}
+	}
+
+	op := &writerOptions{}
+	for _, opt := range opts {
+		opt(op)
+	}
+	w := &Writer{
+		cfg:       cfg,
+		op:        op,
+		uint64buf: make([]byte, 8),
+	}
+
+	if cfg.S3Storage {
+		s3storage, err := common.InitS3storage(ctx, *cfg.S3URI)
+		if err != nil {
+			log.Panic("initS3storage fail",
+				zap.Error(err),
+				zap.String("change feed", cfg.ChangeFeedID))
+		}
+		w.storage = s3storage
+	}
+
+	w.state.Store(started)
+	go w.runFlushToDisk(ctx, cfg.FlushIntervalInMs)
+
+	return w
+}
+
+func (w *Writer) runFlushToDisk(ctx context.Context, flushIntervalInMs int64) {
+	ticker := time.NewTicker(time.Duration(flushIntervalInMs) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if !w.IsRunning() {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Info("runFlushToDisk got canceled", zap.Error(ctx.Err()))
+			return
+		case <-ticker.C:
+			err := w.Flush()
+			if err != nil {
+				log.Error("redo log flush error", zap.Error(err))
+			}
+		}
+	}
+}
+
+// Write ...
+// TODO: more general api with fileName generated by caller
+func (w *Writer) Write(rawData []byte) (int, error) {
+	w.Lock()
+	defer w.Unlock()
+
+	writeLen := int64(len(rawData))
+	if writeLen > w.cfg.MaxLogSize {
+		return 0, errors.Errorf("rawData %d exceeds maximum file size %d", writeLen, w.cfg.MaxLogSize)
+	}
+
+	if w.file == nil {
+		if err := w.openOrNew(len(rawData)); err != nil {
+			return 0, err
+		}
+	}
+
+	if w.size+writeLen > w.cfg.MaxLogSize {
+		if err := w.rotate(); err != nil {
+			return 0, err
+		}
+	}
+	if w.maxCommitTS.Load() < w.eventCommitTS.Load() {
+		w.maxCommitTS.Store(w.eventCommitTS.Load())
+	}
+	// ref: https://github.com/etcd-io/etcd/pull/5250
+	lenField, padBytes := encodeFrameSize(len(rawData))
+	if err := w.writeUint64(lenField, w.uint64buf); err != nil {
+		return 0, err
+	}
+
+	if padBytes != 0 {
+		rawData = append(rawData, make([]byte, padBytes)...)
+	}
+
+	n, err := w.bw.Write(rawData)
```

Comment:
Question: we don't split rawData , does this mean if the length of `rawData` is larger than `w.cfg.MaxLogSize`, a file size could be larger than max log size.

Reply:
could be
```
if int64(len(rawData)) > w.cfg.MaxLogSize {
		return 0, cerror.ErrFileSizeExceed.GenWithStackByArgs(writeLen, w.cfg.MaxLogSize)
	}
```
after pading
the maxlen=len(rawData+padBytes), since  `rawData= append(rawData, make([]byte, padBytes)...)`